Lib/test/test_types.py

```diff
@@ -711,6 +711,18 @@ def test_hash(self):
         self.assertEqual(hash(int | str), hash(str | int))
         self.assertEqual(hash(int | str), hash(typing.Union[int, str]))
 
+    def test_union_of_unhashable(self):
+        class UnhashableMeta(type):
+            __hash__ = None
+
+        class A(metaclass=UnhashableMeta): ...
+        class B(metaclass=UnhashableMeta): ...
+
+        self.assertEqual((A | B).__args__, (A, B))
+        self.assertRaises(TypeError, hash, A | B)
+        self.assertRaises(TypeError, hash, int | B)
+        self.assertRaises(TypeError, hash, A | int)
```

Comment:
Make sure it's the `hash()` call that's causing the TypeError to be raised here, not the union operation itself:

```suggestion
        union1 = A | B
        with self.assertRaises(TypeError):
            hash(union1)

        union2 = int | B
        with self.assertRaises(TypeError):
            hash(union2)

        union3 = A | int
        with self.assertRaises(TypeError):
            hash(union3)
```